synapse/http/server.py

```diff
@@ -260,7 +262,11 @@ def _async_render(self, request, request_metrics):
                 for name, value in m.groupdict().items()
             })
 
-            callback_return = yield callback(request, **kwargs)
+            try:
+                callback_return = yield callback(request, **kwargs)
+            except InteractiveAuthIncompleteError as e:
+                callback_return = (401, e.result)
```

Comment:
This feels like the wrong place to do this check TBH. I guess maybe somewhere like https://github.com/matrix-org/synapse/blob/master/synapse/http/server.py#L119 would be better, but I'm worried we're merging application handling code into the HTTP rest layer

Reply:
I had the same thought tbh; otoh the handlers got so much nicer when I was able to factor this out.

line 119 was my first thought, but it's even further away from the handler.

I guess I could define an `@ui_auth_handler` decorator to wrap the `on_POST` methods?

Reply:
`@ui_auth_handler` could work. I assume that we'd notice it pretty quickly if we forgot to put it on there?

Reply:
>  I assume that we'd notice it pretty quickly if we forgot to put it on there?

well, you'd get a 500 when you expected to get a 401 (all UI auth flows begin with the 401 normally), so I'd say so.

Reply:
Cool. None of these solutions make me jump up and down with joy, but I think erring on the side of explicit wrapper might be the right/safe call here. Am happy to concede though if it feels a bit awful in practice.

Reply:
Ok. PTAL.